compiler/src/iree/compiler/Codegen/Dialect/MicroKernelOps.cpp

```diff
@@ -0,0 +1,324 @@
+        return success();
+      })
+      .Default([&](Type t) { return failure(); });
+}
+
+/// Map `operand` of a `generic_micro_kernel` operation to the operand(s) of
+/// the function call it lowers to.
+static LogicalResult lowerToCallOperands(Location loc, RewriterBase &rewriter,
+                                         Value operand,
+                                         SmallVector<Value> &callOperands) {
+  return TypeSwitch<Type, LogicalResult>(operand.getType())
+      .Case<FloatType, IndexType, IntegerType>([&](auto scalarType) {
+        callOperands.push_back(operand);
+        return success();
+      })
+      .Case<MemRefType>([&](MemRefType memrefType) {
+        auto extractStridedMetadataOp =
+            rewriter.create<memref::ExtractStridedMetadataOp>(loc, operand);
+        // Base ptr.
+        callOperands.push_back(extractStridedMetadataOp.getBaseBuffer());
+        if (memrefType.getRank() == 0) {
+          return success();
+        }
+        // Offset.
+        callOperands.push_back(extractStridedMetadataOp.getOffset());
+        // Strides.
+        for (auto stride : extractStridedMetadataOp.getStrides().drop_back()) {
+          callOperands.push_back(stride);
+        }
+        return success();
+      })
+      .Default([](Type) { return failure(); });
+}
+
+FailureOr<func::CallOp> GenericMicroKernelOp::lowerToFunctionCall(
+    RewriterBase &rewriter) {
+  // TODO handle op with return values if they are scalar.
+  if (getNumResults() != 0) {
+    return rewriter.notifyMatchFailure(
+        getOperation(), "cannot lower to function call operation with results");
+  }
+  // Create the function type based on the operands.
+  SmallVector<Type> callArgumentTypes;
+  for (auto microKernelOpOperandType : getOperation()->getOperandTypes()) {
+    if (failed(getCallArgumentTypeForOperand(rewriter.getContext(),
+                                             microKernelOpOperandType,
+                                             callArgumentTypes))) {
+      return rewriter.notifyMatchFailure(
+          getOperation(), llvm::formatv("failed to lower operand type {0}",
+                                        microKernelOpOperandType));
+    }
+  }
+  SmallVector<Value> callOperands;
+  Location loc = getLoc();
+  for (auto operand : getOperands()) {
+    if (failed(lowerToCallOperands(loc, rewriter, operand, callOperands))) {
+      return rewriter.notifyMatchFailure(
+          getOperation(), "failed to lower operands to function call operands");
+    }
+  }
+  return createFunctionCall(rewriter, getOperation(), getMicroKernelFnName(),
+                            callArgumentTypes, /*callReturnTypes=*/TypeRange{},
+                            callOperands);
+}
+
+//===---------------------------------------------------------------------===//
+// MMT4DMicroKernelOp
+//===---------------------------------------------------------------------===//
```

Comment:
The following ~ 130 lines of C++ code look like they will need to be replicated/adapted for every non-generic microkernel. My only high level comment on this PR then would be, should we get in a room with @benvanik and try to minimize that amount of per-ukernel code --- either by looking for ways to rely more on generic ukernel ops, or by having custom ukernel ops provide more information allowing them to be handled by a single lower-to-function-call implementation? The baseline that this is replacing is https://sourcegraph.com/github.com/iree-org/iree@1ae012ec6c8c81eafb723e08b2aa0d5fc8115dec/-/blob/compiler/src/iree/compiler/Dialect/VM/Conversion/ImportUtils.h?L55-145

Reply:
```either by looking for ways to rely more on generic ukernel ops, or by having custom ukernel ops provide more information allowing them to be handled by a single lower-to-function-call implementation```

Trying to understand the suggestion here. That is kind of the intent of having the `generic_micro_kernel` op. If the actual micro kernel follows the ABI that is captured here, then there is nothing more to do. If you want a specific ABI, then you have a specific micro kernel operation. Are you suggesting some "non-compiler" interface that the actual micro kernel function itself provides some information about the ABI that is picked up and implemented by the lowering to function call. We could do that as a follow up, if something like that is possible... But, for now, I am happy to just write C++ :)

Reply:
I'm trying to see if we can avoid having to write 130 lines of custom C++ code for each ukernel's lowering to a function call. The existing code link worked generically for all ukernels. I'm OK to adjust ukernel interfaces, or write some more tablegen boilerplater, whatever --- is there a way to avoid 130 lines of C++ per ukernels? I think the above link to the existing rewriteToCall code, provides some kind of existence proof here.

Reply:
Maybe look into each of the things that these 130 lines of code do. For example, they bridge over the gap from specifying lhs/rhs/acc tensor, to specifying M/N/K/M0/N0/K0 integers. So maybe we can reduce the boilerplate C++ code to just one function returning a `SmallVector<Value>` giving the arguments list to pass to the function call?

Reply:
I dont want to common out too much.. It just adds coupling that might not be worth it... If we want to avoid writing the 130 C++ lines, then we just need to use `GenericMicroKernelOp` for all micro kernel hand offs. If we want special ABI handling, then we need to write extra C++.

Reply:
Fair enough!